src/scripts/audit/tool.sh

```diff
@@ -0,0 +1,305 @@
+  fi
+
+  # Read artifacts file
+  IFS=$'\n' read -d '' -r -a artifacts < "$artifactsFile"
+  artifactsFile=${artifacts[@]}
+
+  # Read image to node mapping from kubernetes file in nodeToImageMapping array
+  if [ -n "${kubernetesFile}" ]; then
+    getImagesInNodeFile
+  fi
+
+}
+
+function validateAvailableImages {
+
+  info "Validating images available for scanning on this machine..."
+  getMissingImagesList
+  missingImageCount=$?
+  if [[ $missingImageCount -gt 0 ]]; then
+    fail "Found $missingImageCount images missing on current machine. The list of missing images is saved in ${missingImagesReport} file, these images must be pulled to current machine before the script can continue. Use '-d' option to disable image validation."
+  fi
+}
+
+#
+# Initialize workspace dir and generate report header
+#
+function initWorkDir {
+  if [ -z ${workDir} ]; then
+    workDir="./work"
+  fi
+  rm -fr "${workDir}"
+  mkdir -p "${workDir}"
+  missingImagesReport="${workDir}/missing-images.txt"
+  reportFile="${workDir}/report.txt"
+  logFile=${workDir}/tool.log
+  echo "Log for this run can be found in ${logFile}"
+  info "Image scan started at `date -u '+%Y-%m-%d %H.%M.%SUTC'`"
+}
+
+#
+# Read image list, verify hash and get node names (if node info provided) and process
+#
+function scanAndSearchImageList {
+
+  for image in ${imageFile}
+  do
+    imageRepoTag=$(echo $image | cut -d'@' -f1)
+    imageHash=$(echo $image | cut -d'@' -f2)
+    imageId=$(echo $image | cut -d'@' -f3)
+    repository=$(echo $imageRepoTag | cut -d':' -f1)
+    tag=$(echo $imageRepoTag | cut -d':' -f2)
+
+    if [ -z ${imageId} ]; then
+      imageId=$(getImageId "${imageRepoTag}")
+    fi
+
+    imageSize=`docker inspect ${imageId} | jq .[].Size`
+    verifyHashAndGetNodeNames ${repository} ${imageHash} ${imageSize} resultNodeName
+    if [[ -n "${kubernetesFile}" && "${resultNodeName}" == "FailedNotAvailable" ]]; then
+        continue;
+    fi
+
+    if [ -n "$imageId" ]; then
+      info "Scanning image ${imageRepoTag} with id $imageId "
+      scanAndSearchImage "${imageId}" "${imageDir}" "${resultNodeName}" 
+    else
+      fail "Unable to determine image id for image ${image}, image must be pulled before the script can continue"
+    fi
+  done
+}
+
+#
+# Scan image and search for artifact
+#
+function scanAndSearchImage {
+  local imageId=$1
+  local imageDir=$2
+  local nodeNames=$3
+
+  imageDir="${workDir}/${imageId}"
+  mkdir -p "${imageDir}"
+
+  saveImageAndExtractLayers "${imageId}" "${imageDir}"
+  searchArtifactInImage "${imageId}" "${imageDir}" "${nodeNames}"
+  cleanUpImage "${imageDir}"
+}
+
+#
+# Verify image hash and get node names using Kubernetes Node information file
+#
+function verifyHashAndGetNodeNames {
+  local imageRepo=$1
+  local imageHash=$2
+  local imageSize=$3
+  local __result=$4
+
+  #sameSize=false
+  hashMatchFailed=false
+  imageNode=()
+  if [ -z "${kubernetesFile}" ]; then
+    eval $__result="FailedNotAvailable"
```

Comment:
Okay, what does this mean? (This is just the limit of my Bash knowledge)

Reply:
This function gets the information about k8s nodes where a particular image has been pulled. It also verifies that hash value of image on current machine matches with hash value of image in node information file. If the Kubernetes node information file (file containing `kubectl get node` output) is not provided then I skip the hash verification and logic to get node information. Above line (line 218) returns a value indicating that either node information file is not provided or hash verification has failed.  In case of "hash verification failure", script will also print a warning message that "hash verification failed for a particular image" and will continue scanning other images. I'm also not an expert on bash scripting and looked up different ways to return value from function. Above approach of returning values is from this article (3rd approach) https://www.linuxjournal.com/content/return-values-bash-functions . I preferred this approach over "echoing" return value  as it allows to echo warning message in addition to retuning failure. Please let me know if this is not clear or need to be refactored. Thanks.

Reply:
I realize that it's hard to read bash function, below is the pseudo code for this function. I have also refactored the code a little bit to remove above confusing check. Please let me know if you have any suggestions to improve or if something is missing.

```
function verifyHashAndGetNodeNames (Argument imageRepository, Argument imageHash, Argument imageSize, Argument result ){ 

    PreConditions:
    Array "nodeToImageMapping" is populated in initialize method by calling "getImagesInNodeFile" function.
	This array contains Node to Image mapping (i.e. what images are pulled on which k8s nodes)
	It is populated by reading 'k8s node info' file.
	
	for each NodeAndImage pair in "nodeToImageMapping" array:
	  1. Parse NodeName, Image string and Image Size
	  2. Parse Image repo, tag and hash value from Image string obtained in step 1.
	  
	  If (Image repo in step 2 matches with input imageRepository) 
	    a) Check if image size from step 1 matches with imageSize input parameter
		   If match fails and size verification is enabled then provide a warning message and break from loop
		b) Add node name to the array of node-names where image is pulled.
		c) Check if hash value from step 2 matches with input imageHash value
		   If match fails then provide a warning message, set a flag and break from loop
	
    If (lengh of node array is 0 or hash match failed)
        return error value "NotAvailableOrFailed"
    else 
        return array containing node names where image is pulled.

}		
```